File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.js

class Solution {
	/**
	 * @param {number[]} candidates
	 * @param {number} target
	 * @return {number[][]}
	 */
	combinationSum2(candidates, target) {
		let answer = [];
		let buffer = [];
		candidates.sort((a, b) => a - b);
		function backtrack(curTotal = 0, index = 0) {
			if (curTotal === target) {
				answer.push([...buffer]);
				return;
			}
			if (curTotal > target || index >= candidates.length) return;
			buffer.push(candidates[index]);
			backtrack(curTotal + candidates[index], index + 1);
			buffer.pop();
			while (index < candidates.length - 1 && candidates[index] === candidates[index + 1]) index++;
			backtrack(curTotal, index + 1);
		}
		backtrack();
		return answer;
	}

}

File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        subset = []

        def backtrack(i: int = 0, total: int = 0):
            if i >= len(nums) or total > target:
                return
            elif total == target:
                ans.append(subset.copy())
                return
            subset.append(nums[i])
            backtrack(i, total + nums[i])
            subset.pop()
            backtrack(i + 1, total)

        backtrack()

        return ans
